### PR TITLE
RACL gold transmission

### PR DESCRIPTION
**Pull request by pmhalvor**
My implementation of RACL previously did not support gold transmission. 

Now, transmission is applied to the values of the sentiment context attention block, via expression logits. 
These values are defined as a sum between polarity convolutional output and target at expression attention. 
This slightly differs from the original implementation, where these are added to an internal state during attention block. 
However, to keep this code as close to PyTorch as possible, we use this information in the values. The attention should still pick up on the signal correctly, since its a cross-multiplication, and the logits are just emphasizing a transformation of the same elements.  

Query is the output from the shared layers, while keys are the outputs from the polarity convolution. 

Still a little confusion around L2-normalized attention inputs. 
Should maybe take a deeper look to check if these increase or decrease performance, along with where they should and should not be used.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -1785,6 +1784,10 @@ def forward(self, batch):
 
         input_ids = batch[0].to(torch.device(self.device))
         attention_mask = batch[1].to(torch.device(self.device))
+        true_labels = {
+            "expression": batch[2].to(torch.device(self.device)),
+            "target": batch[5].to(torch.device(self.device))
+        }
```
Comment (pmhalvor) [inline src/model.py:1790]: Really only need expression, but added target for easier testing other transmission types

File: src/model.py [hunk 2]
```diff
@@ -1809,62 +1812,66 @@ def forward(self, batch):
...
+            key_target = torch.nn.functional.normalize(target_cnn, p=2, dim=-1).permute(2, 0, 1)
+            value_target = target_cnn.permute(2, 0, 1)
+            mask = ((batch[1]*-1)+1).bool().to(torch.device(self.device))
+
+            target_attn, _ = self.components["relations"][f"stack_{i}"]["target_at_expression"](
+                query=query_expression,
+                key=key_target,
+                value=value_target,
+                key_padding_mask=mask,
+                need_weights=False
```
Comment (pmhalvor) [inline src/model.py:1828]: Cleaner variable names for easier human readability

File: src/model.py [hunk 3]
```diff
@@ -1809,62 +1812,66 @@ def forward(self, batch):
...
-                )
-            else:
-                expression_propagate = expression_attn
 
-            # polarity convolution
+            #### Polarity convolution
             polarity_cnn = self.components["polarity"]["cnn"](polarity_inputs[-1])
 
+            # [sequence, batch, cnn_dim]
             # TODO it looks like both target and expression should each be applied to polarity somehow..?
```
Comment (pmhalvor) [inline src/model.py:1857]: No longer relevant. Delete

File: src/model.py [hunk 4]
```diff
@@ -1912,6 +1919,24 @@ def get_confidence(expression_logits, batch, current_epoch):
...
+        # decide how much true labels should influence attention based on current epoch
+        gold_influence = self.get_prob(self.current_epoch, self.find("warm_up_constant", default=5))
+
+        focus_scope = (  # strengthen signal from true logits
+            gold_influence*true.bool().float()
+                + (1-gold_influence)*logits.argmax(-1)
+        ).detach().to(torch.device(self.device))
+        focus_scope.requires_grad = True
+        
+        return focus_scope  # shape: [batch, sequence]
```
Comment (pmhalvor) [inline src/model.py:1936]: Copied directly from FgFlex implementation. 

Should also update FgFlex to use logits, not cnn outputs, since logits will the entire subtask's information.

File: src/model.py [hunk 5]
```diff
@@ -2215,7 +2240,7 @@ def forward(self, batch):
                     first_transmission = first_transmission.permute(1, 0).unsqueeze(-1).expand(key.shape)
                     # second_transmission = second_transmission.permute(1, 0).unsqueeze(-1).expand(query.shape)
 
-                    # NOTE only apply gold transimssion to keys (first task), so values help "remap" to previous state
+                    # NOTE only apply gold transmission to keys (first task), so values help "remap" to previous state
                     # query = query * second_transmission
                     key = key * first_transmission
                     # value = value * first_transmission
```
Comment (pmhalvor) [inline src/model.py:2246]: This needs to be updated in a new PR
